### PR TITLE
Add `RenderTarget::TextureView`

### PR DESCRIPTION
**Pull request by mrchantey**
# Objective

We can currently set `camera.target` to either an `Image` or `Window`. For OpenXR & WebXR we need to be able to render to a `TextureView`.

This partially addresses #115 as with the addition we can create internal and external xr crates.

## Solution

A `TextureView` item is added to the `RenderTarget` enum. It holds an id which is looked up by a `ManualTextureViews` resource, much like how `Assets<Image>` works.
I believe this approach was first used by @kcking in their [xr fork](https://github.com/kcking/bevy/blob/eb39afd51bcbab38de6efbeeb0646e01e2ce4766/crates/bevy_render/src/camera/camera.rs#L322). The only change is that a `u32` is used to index the textures as `FromReflect` does not support `uuid` and I don't know how to implement that.

---

### Added
Render: Added `RenderTarget::TextureView` as a `camera.target` option, enabling rendering directly to a `TextureView`.

## Migration Guide

References to the `RenderTarget` enum will need to handle the additional field, ie in `match` statements.

---

## Comments
- The [wgpu work](https://github.com/gfx-rs/wgpu/commit/c039a74884a00ba744d9ebccb80e45a2e9e8f80e) done by @expenses allows us to create framebuffer texture views from `wgpu v0.15, bevy 0.10`.
- I got the WebXR techniques from the [xr fork](https://github.com/dekuraan/xr-bevy) by @dekuraan.
- I have tested this with a wip [external webxr crate](https://github.com/mrchantey/forky/blob/018e22bb06b7542419db95f5332c7684931c9c95/crates/bevy_webxr/src/bevy_utils/xr_render.rs#L50) on an Oculus Quest 2. 

Screenshot 2023-03-11 230651

### CODE REVIEW

**Comment by MalekiRe** [comment]
Bumping

File: crates/bevy_render/src/camera/camera.rs [hunk 1]
```diff
@@ -368,6 +370,24 @@ impl CameraRenderGraph {
     }
 }
 
+/// Stores Texture Views used as render targets.
+#[derive(Default, Clone, Resource, ExtractResource)]
+pub struct ManualTextureViews(HashMap<u32, (TextureView, UVec2)>);
```
Comment (IceSentry) [inline crates/bevy_render/src/camera/camera.rs:375]: I'm not sure what "Manual" is supposed to mean here.

I think I'd prefer using a wrapper type for the u32. It wasn't immediately obvious to me what the u32 was supposed to be.

File: crates/bevy_render/src/camera/camera.rs [hunk 2]
```diff
@@ -502,13 +542,16 @@ impl NormalizedRenderTarget {
 /// [`OrthographicProjection`]: crate::camera::OrthographicProjection
 /// [`PerspectiveProjection`]: crate::camera::PerspectiveProjection
 /// [`Projection`]: crate::camera::Projection
+/// [`CoreSet::PostUpdate`]: bevy_app::CoreSet::PostUpdate
```
Comment (IceSentry) [inline crates/bevy_render/src/camera/camera.rs:545]: Why was this line added?

File: crates/bevy_render/src/camera/camera.rs (hunk 1 above)
Comment (IceSentry) [reply crates/bevy_render/src/camera/camera.rs:375]: I'd also add a note in the comment that the UVec2 is the size. Or maybe just say that it stores the texture view and it's size

File: crates/bevy_render/src/camera/camera.rs [hunk 3]
```diff
@@ -436,13 +465,15 @@ impl NormalizedRenderTarget {
             NormalizedRenderTarget::Image(image_handle) => {
                 images.get(image_handle).map(|image| image.texture_format)
             }
+            NormalizedRenderTarget::TextureView(_) => Some(TextureFormat::bevy_default()),
```
Comment (IceSentry) [inline crates/bevy_render/src/camera/camera.rs:468]: I don't think this line is correct. As far as I know, it's possible for a TextureView to be a different format.

File: crates/bevy_render/src/camera/camera.rs (hunk 2 above)
Comment (mrchantey) [reply crates/bevy_render/src/camera/camera.rs]: Git blame is showing this as from an older pr, i guess it has been removed, i've deleted it from this pr.
Screenshot 2023-05-24 144633

File: crates/bevy_render/src/camera/camera.rs (hunk 3 above)
Comment (mrchantey) [reply crates/bevy_render/src/camera/camera.rs:468]: Ok we should check that, my understanding was that the `TextureView` is opaque and we dont have access to the format, @kcking do you have thoughts on this?

File: crates/bevy_render/src/camera/camera.rs (hunk 1 above)
Comment (mrchantey) [reply crates/bevy_render/src/camera/camera.rs:375]: Ok great I've updated the comment.

`Manual` is intended to describe that these are texture views we are managing ourselves outside of Bevy, either by creating them in wgpu or being given them by OpenXR, WebXR etc. I agree its not obvious, any suggestions?

Comment (mrchantey) [reply crates/bevy_render/src/camera/camera.rs]: I've added the `ManualTextureViewHandle` wrapper type, im a little concerned its bloating the `camera.rs` file, would you like the `ManualTextureView` struct moved to a seperate file?

File: crates/bevy_render/src/camera/camera.rs (hunk 3 above)
Comment (mrchantey) [reply crates/bevy_render/src/camera/camera.rs:486]: if it is something we want to track, maybe instead of storing a `(TextureView,UVec2)` tuple, we can store some struct

```rs
pub struct ManualTextureView {
    texture_view: TextureView,
    size: UVec2,
    format: TextureFormat,
}
```

Comment (kcking) [reply crates/bevy_render/src/camera/camera.rs:486]: Agreed, storing the correct `TextureFormat` makes the most sense to me.

File: crates/bevy_render/src/camera/camera.rs (hunk 1 above)
Comment (mrchantey) [reply crates/bevy_render/src/camera/camera.rs]: I've created `manual_texture_view.rs`, let me know if you want that reverted.

**Comment by IceSentry** [comment]
So, the implementation seems good to me, but I'm not sure I understand _why_ this is necessary. Isn't the Image render target just a wrapper around targetting a TextureView? As in, couldn't you just use the render to image feature and use the texture view of the image?

**Comment by kcking** [comment]
> So, the implementation seems good to me, but I'm not sure I understand _why_ this is necessary. Isn't the Image render target just a wrapper around targetting a TextureView? As in, couldn't you just use the render to image feature and use the texture view of the image?

There's a quick [discord conversation](https://discord.com/channels/691052431525675048/829769375618826251/1104537972902203513) about this, essentially OpenXR requires you to use its `TextureView`s, you can't provider your own. AFAIU bevy's render to image feature creates new `TextureView`s so it's incompatible with OpenXR

**Comment by IceSentry** [comment]
Alright, makes sense. In that case this PR seems fine.

**Review by IceSentry — APPROVED** [review]

File: crates/bevy_render/src/camera/camera.rs [hunk 4]
```diff
@@ -376,6 +376,8 @@ pub enum RenderTarget {
     Window(WindowRef),
     /// Image to which the camera's view is rendered.
     Image(Handle<Image>),
+    /// Texture View to which the camera's view is rendered.
+    TextureView(ManualTextureViewHandle),
```
Comment (paul-hansen) [inline crates/bevy_render/src/camera/camera.rs:380]: Might be worth adding a doc comment about the use case since it's been asked a few times now and can be useful context for why you would use one over the other.

```suggestion
    /// Texture View to which the camera's view is rendered.
    /// Useful when the texture view needs to be created outside of Bevy, for example OpenXR. 
    TextureView(ManualTextureViewHandle),
```

File: crates/bevy_render/src/camera/camera.rs [hunk 5]
```diff
@@ -387,6 +389,8 @@ pub enum NormalizedRenderTarget {
     Window(NormalizedWindowRef),
     /// Image to which the camera's view is rendered.
     Image(Handle<Image>),
+    /// Texture View to which the camera's view is rendered.
+    TextureView(ManualTextureViewHandle),
```
Comment (paul-hansen) [inline crates/bevy_render/src/camera/camera.rs:393]: Same as my other suggestion
```suggestion
    /// Texture View to which the camera's view is rendered.
    /// Useful when the texture view needs to be created outside of Bevy, for example OpenXR. 
    TextureView(ManualTextureViewHandle),
```

**Review by paul-hansen — APPROVED** [review]
Added a suggestion for adding a comment regarding the use case for this. 
Looks good otherwise, though I'm not a rendering expert. Has my approval with or without my suggestion.

File: crates/bevy_render/src/camera/camera.rs [hunk 6]
```diff
@@ -387,6 +390,9 @@ pub enum NormalizedRenderTarget {
     Window(NormalizedWindowRef),
     /// Image to which the camera's view is rendered.
     Image(Handle<Image>),
+    /// Texture View to which the camera's view is rendered.
+    /// Useful when the texture view needs to be created outside of Bevy, for example OpenXR.
```
Comment (paul-hansen) [inline crates/bevy_render/src/camera/camera.rs:394]:
```suggestion
    /// Useful when the texture view needs to be created outside of Bevy, for example OpenXR.
```

File: crates/bevy_render/src/camera/camera.rs [hunk 7]
```diff
@@ -376,6 +376,9 @@ pub enum RenderTarget {
     Window(WindowRef),
     /// Image to which the camera's view is rendered.
     Image(Handle<Image>),
+    /// Texture View to which the camera's view is rendered.
+    /// Useful when the texture view needs to be created outside of Bevy, for example OpenXR.
```
Comment (paul-hansen) [inline crates/bevy_render/src/camera/camera.rs:380]:
```suggestion
    /// Useful when the texture view needs to be created outside of Bevy, for example OpenXR.
```

**Review by paul-hansen — COMMENTED** [review]
Oof looks like my suggestion somehow ended up with trailing spaces and it broke cargo format. My bad.

**Comment by alice-i-cecile** [comment]
Straightforward enough, and on the critical path. Let's see what the community can do with this <3
